### PR TITLE
MAHOUT-1817  Implement caching in Flink Bindings

### PR DESCRIPTION
**Pull request by andrewpalumbo**
As a temporary measure, use this method to persist the `DataSet` to the filesystem when caching rather that drmDfsRead()/Write.

Todo:
 
1.  ~~Break up into `persist` and `readPersistedDataset` methods and only read a persisted dataset if it is already cached.~~
2.     ~~Use a property setting for the base dir~~.
3.     ~~Check to make sure that this method maintains parallelism deg for the dataset, if not set the new parallelism degree to the original~~

### CODE REVIEW

**Comment by smarthi** [comment]
lgtm

File: flink/src/main/scala/org/apache/mahout/flinkbindings/drm/CheckpointedFlinkDrm.scala [hunk 1]
```diff
@@ -76,20 +94,31 @@ class CheckpointedFlinkDrm[K: ClassTag:TypeInformation](val ds: DrmDataSet[K],
...
     }
-    implicit val typeInformation = createTypeInformation[(K,Vector)]
+    val _ds = readPersistedDataSet(cacheFileName, ds)
 
-    val _ds = persist(ds, persistanceRootDir + cacheFileName)
+    // We may want to look more closely at this:
+    // since we've cached a drm, triggering a computation
+    // it may not make sense to keep the same parallelism degree
+    if (!(parallelismDeg == _ds.getParallelism)) {
+      _ds.setParallelism(parallelismDeg).rebalance()
```
Comment (andrewpalumbo) [inline flink/src/main/scala/org/apache/mahout/flinkbindings/drm/CheckpointedFlinkDrm.scala]: not sure what to do here, this `rebalance` call seems to adding some time to the caching call.

Comment (smarthi) [inline flink/src/main/scala/org/apache/mahout/flinkbindings/drm/CheckpointedFlinkDrm.scala]: Is it needed? IIRC it was added when we had skewed partitions in previous flink releases.

Comment (andrewpalumbo) [inline flink/src/main/scala/org/apache/mahout/flinkbindings/drm/CheckpointedFlinkDrm.scala]: I guess my real question is whether or not to check and re-set the parallelism after  reading in from the cache:
```if (!(parallelismDeg == _ds.getParallelism)) {
      _ds.setParallelism(parallelismDeg).rebalance()
}
```

several operations may result in a much smaller matrix than the original DRM but it is not calculated until here.  Its really a case by case thing, so difficult to figure out.

Comment (andrewpalumbo) [inline flink/src/main/scala/org/apache/mahout/flinkbindings/drm/CheckpointedFlinkDrm.scala]: yes @smarthi  so I will I will likely remove this check tomorrow and then push the rest as is.   I guess we need to go through and make sure that all operations are setting the parallelism if they need to.

File: flink/src/main/scala/org/apache/mahout/flinkbindings/drm/CheckpointedFlinkDrm.scala [hunk 2]
```diff
@@ -76,20 +94,38 @@ class CheckpointedFlinkDrm[K: ClassTag:TypeInformation](val ds: DrmDataSet[K],
...
+      * final drm after computation is actually triggered
+      *
+      *  // We may want to look more closely at this:
+      *  // since we've cached a drm, triggering a computation
+      *  // it may not make sense to keep the same parallelism degree
+      *  if (!(parallelismDeg == _ds.getParallelism)) {
+      *    _ds.setParallelism(parallelismDeg).rebalance()
+      *  }
+      *
+      */
```
Comment (andrewpalumbo) [inline flink/src/main/scala/org/apache/mahout/flinkbindings/drm/CheckpointedFlinkDrm.scala]: commented out getParallelism setting in cache() for now.  will push soon if no objections.

Comment (smarthi) [inline flink/src/main/scala/org/apache/mahout/flinkbindings/drm/CheckpointedFlinkDrm.scala]: In light of MAHOUT-1819, wherein its been agreed that parallelism would only be set inside MahoutFlinkContext for the ExecutionEnvironment, I think this jira can be safely marked as 'Resolved'.

**Pull request merged**